compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64ExtendingReadNode.java

```diff
@@ -93,7 +100,13 @@ public static void replace(ReadNode readNode) {
         BarrierType barrierType = readNode.getBarrierType();
         boolean nullCheck = readNode.getNullCheck();
         FrameState stateBefore = readNode.stateBefore();
-        AArch64ReadNode clone = new AArch64ReadNode(address, location, stamp, guard, barrierType, nullCheck, stateBefore, accessStamp, isSigned);
+        AArch64ExtendingReadNode clone;
+        boolean isVolatile = (readNode instanceof AArch64VolatileReadNode);
+        if (isVolatile) {
+            clone = new AArch64VolatileExtendingReadNode(address, location, stamp, guard, barrierType, nullCheck, stateBefore, accessStamp, isSigned);
```

Comment:
Will this branch be true sometimes? I checked the `emitCode()` of  `MemoryOp`, and it seems that it will not generate `ldar` instruction when `isVolatile` is true. So I think it may cause one issue that we elided the membar before but didn't use the `ldar` instruction. So I think when `isVolatile==true` set up, we should just return or do nothing, right?

Reply:
I don't completely follow what you are saying here. However, I don't think it is worth you continuing to look at this PR.

Firstly, the patch is on hold while Roland Westrelin is looking at a better solution. It has not been updated for several months because ee are waiting to see if there is a better solution for the following reason:

The current model for volatile reads/writes in the IR graph relies on MemBar nodes to add volatile behaviour to normal Read/Write nodes. The patch retains links between Read/Write nodes that represent volatile operations and the enclosing barriers. A late phase detects these links and replaces the Read.Write nodes with variants which the back end translates to ldar or stlr. The back end also avoids generating code for barriers that are linked to these nodes. The patch was correct when it was last tested but it is fragile. It is very easy for changes in many compiler phases to disconnect the Read/Wrote nodes from the MemBar nodes.

Roland is working on a solution which avoids the use of barriers. Instead a Read/Write node will include internal state and behaviour to record the fact that it is volatile. This requires changing the phases which relocate memory nodes so that they do not propagate nodes past volatile Read/Write nodes (i.e. volatile Read/Write nodes get treated like the current MemBar nodes). This is quite a big change. However, it ought to be less fragile than the current patch.

Secondly, I don't think your comment about isVolatile not being true is correct (I'm not even sure I know what you are pointing at).

The code you highlighted is part of a phase that combines Read nodes with sign extensions. It identifies a Read node that is already known to be volatile, an instance of AArch64VolatileReadNode (which subclasses ReadNode).

Now an AArch64VolatileReadNode will have been introduced by a prior phase that detects links to a pair of MemBar nodes, one that precedes it and one that follows it. So, if the second phase does not apply then the back end will detect an instance of AArch64VolatileReadNode and generate an ldar. It also detects the MemBar linksand avoids generating barriers for the MemBars.

In cases where the 2nd phase does apply it combines the read with a sign extension, relying on the act that AArch64 can automatically zero-extend sub-word loads to a full word. So, it replaces the Read + Extend with a new node that the back end understands. There are two cases:

1. a normal (non-volatile) ReadNode + ExtendNode pair is replaced with an AArch64ExtendingReadNode
2.  an AArch64VolatileReadNode + ExtendNode is replaced with an AArch64VolatileExtendingReadNode

Note that AArch64VolatileExtendingReadNode is a subclasss of AArch64ExtendingReadNode.

The code you linked handles the second case when the original node is an AArch64VolatileReadNode. After that replacement the AArch64VolatileExtendingReadNode is still linked to its preceding and traling MemBar nodes.

That replacement does not stop an ldar being generated in the back end. When the back end detects an instance of AArch64ExtendingReadNode it includes a check to see if the node is an instance of subclass AArch64VolatileExtendingReadNode. In that case it generates an ldar. If it is a plain AArch64ExtendingReadNode it generates an ldr (or maybe lrdh, ldrb or whatever).

Well, at least the patch definitely did that when I last tested it some months ago. I don't see any point checking whether that is still the case until Roland's attempt to provide an alternative solution is completed. If his attempt fails and this patch needs to be picked up again part of the task of updating it will be to check that this mechanism has not been broken by patches that have happened since it was last looked at.

Reply:
Hi Andrew,

I just noticed that `ldar` isn't generated when `isVolatile` is true in `MemoryOp`, which is the op related to `AArch64VolatileExtendingReadNode`. So that's what my points here. Maybe I made some mistakes. If I misunderstood, please ignore. Thanks!

 Also thanks for your explanation about this patch status. I payed attention to this PR just because I also have the same optimization task on AArch64 like this one.  Then I will not continue on it, because I have finished learning about this PR, and also gained a lot. Thank you!

Best Regards,
Xiaohong

Reply:
`@Override
        public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
            if (state != null) {
                crb.recordImplicitException(masm.position(), state);
            }
            AArch64Address address = input.toAddress();
            Register dst = asRegister(result);

            assert (!isVolatile || address.getAddressingMode() == AArch64Address.AddressingMode.BASE_REGISTER_ONLY);

            if (isSigned) {
                masm.ldrs(targetSize, srcSize, dst, address);
            } else {
                masm.ldr(srcSize, dst, address);
            }
        }`

This is the related codes in `MemoryOp` I mentioned above.

Reply:
Ok, I understand what you are saying now. Thank you for providing more explanation.

You are correct that the implementation of AArch64.memoryOp is incomplete. It needs to use ldar when isVolatile is set to true. It also needs to deal with the case where isSigned and isVolatile are both true and targetSize != sourceSize. In that case the result of the ldar wil need to be installed in a new register and then extended as a separate step.